# CPSC 330 - Applied Machine Learning 

## Homework 8: Time series
**Due date: [April 10, 11:59pm](https://github.com/UBC-CS/cpsc330-2023W2?tab=readme-ov-file#deliverable-due-dates-tentative).**

## Imports

In [1]:
from hashlib import sha1

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OrdinalEncoder, OneHotEncoder

from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor

from sklearn.metrics import r2_score

## Submission instructions
<hr>

_points: 2_

You will receive marks for correctly submitting this assignment. To submit this assignment, follow the instructions below:

- **You may work on this assignment in a group (group size <= 4) and submit your assignment as a group.** 
- Below are some instructions on working as a group.  
    - The maximum group size is 4. 
    - You can choose your own group members. 
    - Use group work as an opportunity to collaborate and learn new things from each other. 
    - Be respectful to each other and make sure you understand all the concepts in the assignment well. 
    - It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline. [Here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members) are some instructions on adding group members in Gradescope.  
- Be sure to follow the [homework submission instructions](https://github.com/UBC-CS/cpsc330-2023W2/blob/main/docs/homework_instructions.md).
- Upload the .ipynb file to PrairieLearn.

<br><br>

## Exercise 1: time series prediction

In this exercise we'll be looking at a [dataset of avocado prices](https://www.kaggle.com/neuromusic/avocado-prices). You should start by downloading the dataset. We will be forcasting average avocado price for the next week. 

In [2]:
df = pd.read_csv("data/avocado.csv", parse_dates=["Date"], index_col=0)
df.head()

Date  AveragePrice  Total Volume     4046       4225    4770  \
0 2015-12-27          1.33      64236.62  1036.74   54454.85   48.16   
1 2015-12-20          1.35      54876.98   674.28   44638.81   58.33   
2 2015-12-13          0.93     118220.22   794.70  109149.67  130.50   
3 2015-12-06          1.08      78992.15  1132.00   71976.41   72.58   
4 2015-11-29          1.28      51039.60   941.48   43838.39   75.78   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     8696.87     8603.62       93.25          0.0  conventional  2015  Albany  
1     9505.56     9408.07       97.49          0.0  conventional  2015  Albany  
2     8145.35     8042.21      103.14          0.0  conventional  2015  Albany  
3     5811.16     5677.40      133.76          0.0  conventional  2015  Albany  
4     6183.95     5986.26      197.69          0.0  conventional  2015  Albany

In [3]:
df.shape

(18249, 13)

In [4]:
df["Date"].min()

Timestamp('2015-01-04 00:00:00')

In [5]:
df["Date"].max()

Timestamp('2018-03-25 00:00:00')

It looks like the data ranges from the start of 2015 to March 2018 (~2 years ago), for a total of 3.25 years or so. Let's split the data so that we have a 6 months of test data.

In [6]:
split_date = '20170925'
df_train = df[df["Date"] <= split_date]
df_test  = df[df["Date"] >  split_date]

In [7]:
assert len(df_train) + len(df_test) == len(df)

<br><br>

<!-- BEGIN QUESTION -->

### 1.1 How many time series? 

_Points:_ 4

In the [Rain in Australia](https://www.kaggle.com/datasets/jsphyg/weather-dataset-rattle-package) dataset from lecture, we had different measurements for each Location. How about this avocado sales dataset? For which categorical feature(s), if any, do we have separate measurements? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.1
    
</div>

In [8]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15441 entries, 0 to 50
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Date          15441 non-null  datetime64[ns]
 1   AveragePrice  15441 non-null  float64       
 2   Total Volume  15441 non-null  float64       
 3   4046          15441 non-null  float64       
 4   4225          15441 non-null  float64       
 5   4770          15441 non-null  float64       
 6   Total Bags    15441 non-null  float64       
 7   Small Bags    15441 non-null  float64       
 8   Large Bags    15441 non-null  float64       
 9   XLarge Bags   15441 non-null  float64       
 10  type          15441 non-null  object        
 11  year          15441 non-null  int64         
 12  region        15441 non-null  object        
dtypes: datetime64[ns](1), float64(9), int64(1), object(2)
memory usage: 1.6+ MB


There are 2 categorical features in this dataset: "type" and "region".

In [9]:
df.sort_values(by=["region", "Date"]).head()
    

Date  AveragePrice  Total Volume     4046      4225    4770  \
51 2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
51 2015-01-04          1.79       1373.95    57.42    153.88    0.00   
50 2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
50 2015-01-11          1.77       1182.56    39.00    305.12    0.00   
49 2015-01-18          1.17      44511.28   914.14  31540.32  135.77   

    Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
51     9716.46     9186.93      529.53          0.0  conventional  2015   
51     1162.65     1162.65        0.00          0.0       organic  2015   
50     8424.77     8036.04      388.73          0.0  conventional  2015   
50      838.44      838.44        0.00          0.0       organic  2015   
49    11921.05    11651.09      269.96          0.0  conventional  2015   

    region  
51  Albany  
51  Albany  
50  Albany  
50  Albany  
49  Albany

In [10]:
df.sort_values(by=["type", "Date"]).head()

Date  AveragePrice  Total Volume       4046       4225      4770  \
51 2015-01-04          1.22      40873.28    2819.50   28287.42     49.90   
51 2015-01-04          1.00     435021.49  364302.39   23821.16     82.15   
51 2015-01-04          1.08     788025.06   53987.31  552906.04  39995.03   
51 2015-01-04          1.01      80034.32   44562.12   24964.23   2752.35   
51 2015-01-04          1.02     491738.00    7193.87  396752.18    128.82   

    Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
51     9716.46     9186.93      529.53          0.0  conventional  2015   
51    46815.79    16707.15    30108.64          0.0  conventional  2015   
51   141136.68   137146.07     3990.61          0.0  conventional  2015   
51     7755.62     6064.30     1691.32          0.0  conventional  2015   
51    87663.13    87406.84      256.29          0.0  conventional  2015   

                 region  
51               Albany  
51              Atlanta  
51  BaltimoreWashington  
51                Boise  
51               Boston

There are multiple measurements for both regions and type. When sorting the dataframe by date and region, we see that there are multiple measurements for the same region on the same day. We see that there are also multiple measurements for the same type on the same day when sorting by date and type instead.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.2 Equally spaced measurements? 

_Points:_ 4

In the Rain in Australia dataset, the measurements were generally equally spaced but with some exceptions. How about with this dataset? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.2
    
</div>

In [11]:
result_data = []

for (region, type_), group in df.groupby(['region', 'type']):
    diff_counts = group["Date"].sort_values().diff().value_counts()
    row_data = {"Region": region, "Type": type_}
    for index, value in diff_counts.items():
        row_data["%d days" % index.days] = value
    result_data.append(row_data)

result_df = pd.DataFrame(result_data)
print(result_df)


                  Region          Type  7 days  14 days  21 days
0                 Albany  conventional     168      NaN      NaN
1                 Albany       organic     168      NaN      NaN
2                Atlanta  conventional     168      NaN      NaN
3                Atlanta       organic     168      NaN      NaN
4    BaltimoreWashington  conventional     168      NaN      NaN
..                   ...           ...     ...      ...      ...
103              TotalUS       organic     168      NaN      NaN
104                 West  conventional     168      NaN      NaN
105                 West       organic     168      NaN      NaN
106     WestTexNewMexico  conventional     168      NaN      NaN
107     WestTexNewMexico       organic     163      1.0      1.0

[108 rows x 5 columns]


The results in this dataset are generally equally spaced with measurements at 7 days for all but 1 location. For the region "WestTexNewMexico" with type "organic", there are measurements at 7 days, 14 days, and 21 days.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.3 Interpreting regions 

_Points:_ 4

In the Rain in Australia dataset, each location was a different place in Australia. For this dataset, look at the names of the regions. Do you think the regions are also all distinct, or are there overlapping regions? Justify your answer by referencing the data.

<div class="alert alert-warning">

Solution_1.3
    
</div>

In [12]:
df['region'].unique()

array(['Albany', 'Atlanta', 'BaltimoreWashington', 'Boise', 'Boston',
       'BuffaloRochester', 'California', 'Charlotte', 'Chicago',
       'CincinnatiDayton', 'Columbus', 'DallasFtWorth', 'Denver',
       'Detroit', 'GrandRapids', 'GreatLakes', 'HarrisburgScranton',
       'HartfordSpringfield', 'Houston', 'Indianapolis', 'Jacksonville',
       'LasVegas', 'LosAngeles', 'Louisville', 'MiamiFtLauderdale',
       'Midsouth', 'Nashville', 'NewOrleansMobile', 'NewYork',
       'Northeast', 'NorthernNewEngland', 'Orlando', 'Philadelphia',
       'PhoenixTucson', 'Pittsburgh', 'Plains', 'Portland',
       'RaleighGreensboro', 'RichmondNorfolk', 'Roanoke', 'Sacramento',
       'SanDiego', 'SanFrancisco', 'Seattle', 'SouthCarolina',
       'SouthCentral', 'Southeast', 'Spokane', 'StLouis', 'Syracuse',
       'Tampa', 'TotalUS', 'West', 'WestTexNewMexico'], dtype=object)

There are overlapping regions in this dataset. For example, California is a distinct region in the dataset, but several other regions in the dataset are cities within California (including SanDiego, Sacramento, LosAngeles). There is also a "West" region that may refer to all of the western United States, and a "TotalUS" region which would include all other regions within the dataset.

<!-- END QUESTION -->

<br><br>

We will use the entire dataset despite any location-based weirdness uncovered in the previous part.

We will be trying to forecast the avocado price. The function below is adapted from [Lecture 19](https://github.com/UBC-CS/cpsc330-2023W2/blob/main/lectures/mathias/19_time-series.ipynb), with some improvements.

In [13]:
def create_lag_feature(df, orig_feature, lag, groupby, new_feature_name=None, clip=False):
    """
    Creates a new feature that's a lagged version of an existing one.
    
    NOTE: assumes df is already sorted by the time columns and has unique indices.
    
    Parameters
    ----------
    df : pandas.core.frame.DataFrame
        The dataset.
    orig_feature : str
        The column name of the feature we're copying
    lag : int
        The lag; negative lag means values from the past, positive lag means values from the future
    groupby : list
        Column(s) to group by in case df contains multiple time series
    new_feature_name : str
        Override the default name of the newly created column
    clip : bool
        If True, remove rows with a NaN values for the new feature
    
    Returns
    -------
    pandas.core.frame.DataFrame
        A new dataframe with the additional column added.
        
    """
        
    if new_feature_name is None:
        if lag < 0:
            new_feature_name = "%s_lag%d" % (orig_feature, -lag)
        else:
            new_feature_name = "%s_ahead%d" % (orig_feature, lag)
    
    new_df = df.assign(**{new_feature_name : np.nan})
    for name, group in new_df.groupby(groupby):        
        if lag < 0: # take values from the past
            new_df.loc[group.index[-lag:],new_feature_name] = group.iloc[:lag][orig_feature].values
        else:       # take values from the future
            new_df.loc[group.index[:-lag], new_feature_name] = group.iloc[lag:][orig_feature].values
            
    if clip:
        new_df = new_df.dropna(subset=[new_feature_name])
        
    return new_df

We first sort our dataframe properly:

In [14]:
df_sort = df.sort_values(by=["region", "type", "Date"]).reset_index(drop=True)
df_sort.head()

Date  AveragePrice  Total Volume     4046      4225    4770  \
0 2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1 2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2 2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3 2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4 2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     9716.46     9186.93      529.53          0.0  conventional  2015  Albany  
1     8424.77     8036.04      388.73          0.0  conventional  2015  Albany  
2    11921.05    11651.09      269.96          0.0  conventional  2015  Albany  
3    10845.82    10103.35      742.47          0.0  conventional  2015  Albany  
4     9323.18     9170.82      152.36          0.0  conventional  2015  Albany

We then call `create_lag_feature`. This creates a new column in the dataset `AveragePriceNextWeek`, which is the following week's `AveragePrice`. We have set `clip=True` which means it will remove rows where the target would be missing.

In [15]:
df_hastarget = create_lag_feature(df_sort, "AveragePrice", +1, ["region", "type"], "AveragePriceNextWeek", clip=True)
df_hastarget.head()

Date  AveragePrice  Total Volume     4046      4225    4770  \
0 2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1 2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2 2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3 2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4 2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0     9716.46     9186.93      529.53          0.0  conventional  2015   
1     8424.77     8036.04      388.73          0.0  conventional  2015   
2    11921.05    11651.09      269.96          0.0  conventional  2015   
3    10845.82    10103.35      742.47          0.0  conventional  2015   
4     9323.18     9170.82      152.36          0.0  conventional  2015   

   region  AveragePriceNextWeek  
0  Albany                  1.24  
1  Albany                  1.17  
2  Albany                  1.06  
3  Albany                  0.99  
4  Albany                  0.99

Our goal is to predict `AveragePriceNextWeek`. 

Let's split the data:

In [16]:
df_train = df_hastarget[df_hastarget["Date"] <= split_date]
df_test  = df_hastarget[df_hastarget["Date"] >  split_date]

<br><br>

### 1.4 `AveragePrice` baseline 

_Points:_ 4

Soon we will want to build some models to forecast the average avocado price a week in advance. Before we start with any ML though, let's try a baseline. Previously we used `DummyClassifier` or `DummyRegressor` as a baseline. This time, we'll do something else as a baseline: we'll assume the price stays the same from this week to next week. So, we'll set our prediction of "AveragePriceNextWeek" exactly equal to "AveragePrice", assuming no change. That is kind of like saying, "If it's raining today then I'm guessing it will be raining tomorrow". This simplistic approach will not get a great score but it's a good starting point for reference. If our model does worse that this, it must not be very good. 

Using this baseline approach, what $R^2$ do you get on the train and test data?

<div class="alert alert-warning">

Solution_1.4
    
</div>

In [17]:
train_r2 = r2_score(df_train["AveragePriceNextWeek"], df_train["AveragePrice"])
print("R^2 train: %.5f" % train_r2)

R^2 train: 0.82858


In [18]:
test_r2 = r2_score(df_test["AveragePriceNextWeek"], df_test["AveragePrice"])
print("R^2 test: %.5f" % test_r2)

R^2 test: 0.76318


<br><br>

<!-- BEGIN QUESTION -->

### 1.5 Forecasting average avocado price

_Points:_ 10

Now that the baseline is done, let's build some models to forecast the average avocado price a week later. Experiment with a few approachs for encoding the date. Justify the decisions you make. Which approach worked best? Report your test score and briefly discuss your results.

Benchmark: you should be able to achieve $R^2$ of at least 0.79 on the test set. We got to 0.80, but not beyond that. Let us know if you do better!

Note: because we only have 2 splits here, we need to be a bit wary of overfitting on the test set. Try not to test on it a ridiculous number of times. If you are interested in some proper ways of dealing with this, see for example sklearn's [TimeSeriesSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html), which is like cross-validation for time series data.

<div class="alert alert-warning">

Solution_1.3
    
</div>

In [19]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15441 entries, 0 to 18222
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Date                  15441 non-null  datetime64[ns]
 1   AveragePrice          15441 non-null  float64       
 2   Total Volume          15441 non-null  float64       
 3   4046                  15441 non-null  float64       
 4   4225                  15441 non-null  float64       
 5   4770                  15441 non-null  float64       
 6   Total Bags            15441 non-null  float64       
 7   Small Bags            15441 non-null  float64       
 8   Large Bags            15441 non-null  float64       
 9   XLarge Bags           15441 non-null  float64       
 10  type                  15441 non-null  object        
 11  year                  15441 non-null  int64         
 12  region                15441 non-null  object        
 13  AveragePriceNext

In [24]:
numeric_features = [
    "AveragePrice",
    "Total Volume",
    "4046",
    "4225",
    "4770",
    "Total Bags",
    "Small Bags",
    "Large Bags",
    "XLarge Bags",
    "year",
]
categorical_features = [
    "type",
    "region",
]
drop_features = ["Date"]
target = ["AveragePriceNextWeek"]

In [32]:
# Adapted from Lecture 19
from sklearn.pipeline import make_pipeline
from sklearn.compose import make_column_transformer

def preprocess_features(
    train_df,
    test_df,
    numeric_features,
    categorical_features,
    drop_features,
    target
):

    all_features = set(numeric_features + categorical_features + drop_features + target)
    if set(train_df.columns) != all_features:
        print("Missing columns", set(train_df.columns) - all_features)
        print("Extra columns", all_features - set(train_df.columns))
        raise Exception("Columns do not match")

    numeric_transformer = make_pipeline(
        SimpleImputer(strategy="median"), StandardScaler()
    )
    categorical_transformer = make_pipeline(
        SimpleImputer(strategy="constant", fill_value="missing"),
        OneHotEncoder(handle_unknown="ignore", sparse=False),
    )

    preprocessor = make_column_transformer(
        (numeric_transformer, numeric_features),
        (categorical_transformer, categorical_features),
        ("drop", drop_features),
    )
    preprocessor.fit(train_df)
    ohe_feature_names = (
        preprocessor.named_transformers_["pipeline-2"]
        .named_steps["onehotencoder"]
        .get_feature_names_out(categorical_features)
        .tolist()
    )
    new_columns = numeric_features + ohe_feature_names

    X_train_enc = pd.DataFrame(
        preprocessor.transform(train_df), index=train_df.index, columns=new_columns
    )
    X_test_enc = pd.DataFrame(
        preprocessor.transform(test_df), index=test_df.index, columns=new_columns
    )

    y_train = train_df["AveragePriceNextWeek"]
    y_test = test_df["AveragePriceNextWeek"]

    return X_train_enc, y_train, X_test_enc, y_test, preprocessor




In [47]:
#adapted from lecture 19

def score_lr_print_coeff(preprocessor, train_df, y_train, test_df, y_test, X_train_enc):
    lr_pipe = make_pipeline(preprocessor, Ridge())
    lr_pipe.fit(train_df, y_train)
    print("Train score: {:.2f}".format(lr_pipe.score(train_df, y_train)))
    print("Test score: {:.2f}".format(lr_pipe.score(test_df, y_test)))
    lr_coef = pd.DataFrame(
        data=lr_pipe.named_steps["ridge"].coef_.flatten(),
        index=X_train_enc.columns,
        columns=["Coef"],
    )
    return lr_coef.sort_values(by="Coef", ascending=False)

Method 1) No Date Encoding - default

In [48]:
# Adapted from Lecture 19
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    df_train,
    df_test,
    numeric_features,
    categorical_features,
    drop_features, target
)

score_lr_print_coeff(preprocessor, df_train, y_train, df_test, y_test, X_train_enc)

Train score: 0.85
Test score: 0.79


Coef
AveragePrice                0.328682
region_SanFrancisco         0.087368
region_HartfordSpringfield  0.084933
region_NewYork              0.067025
type_organic                0.049510
...                              ...
region_Denver              -0.045930
type_conventional          -0.049510
region_DallasFtWorth       -0.065811
region_SouthCentral        -0.065815
region_Houston             -0.075620

[66 rows x 1 columns]

Method 2) Encoding time as a number

In [37]:
#adapted from lecture 19
first_day = df_train["Date"].min()

train_df = df_train.assign(
    Days_since=df_train["Date"].apply(lambda x: (x - first_day).days)
)
test_df = df_test.assign(
    Days_since=df_test["Date"].apply(lambda x: (x - first_day).days)
)

train_df.sort_values(by="Date").head()

X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    train_df,
    test_df,
    numeric_features + ["Days_since"],
    categorical_features,
    drop_features,
    target
)

score_lr_print_coeff(preprocessor, train_df, y_train, test_df, y_test, X_train_enc)

Train score: 0.85
Test score: 0.78


Coef
AveragePrice                0.324498
region_SanFrancisco         0.091838
region_HartfordSpringfield  0.089592
region_NewYork              0.070944
region_Philadelphia         0.052258
...                              ...
region_Denver              -0.047782
type_conventional          -0.052177
region_DallasFtWorth       -0.069429
region_SouthCentral        -0.070072
region_Houston             -0.079556

[67 rows x 1 columns]

Method 3) One-hot encoding of the month

In [54]:
train_df = train_df.assign(
    Month=train_df["Date"].apply(lambda x: x.month_name())
)  # x.month_name() to get the actual string
test_df = test_df.assign(Month=test_df["Date"].apply(lambda x: x.month_name()))

train_df[["Date", "Month"]].sort_values(by="Month")

X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    train_df,
    test_df,
    numeric_features + ["Days_since"],
    categorical_features + ["Month"],
    drop_features,
    target
)

score_lr_print_coeff(preprocessor, train_df, y_train, test_df, y_test, X_train_enc)

Train score: 0.85
Test score: 0.80


Coef
AveragePrice         0.316943
year                 0.247612
Month_September      0.112086
Month_October        0.100772
region_SanFrancisco  0.100081
...                       ...
region_Houston      -0.086358
Month_April         -0.087222
Month_February      -0.120834
Month_January       -0.179477
Days_since          -0.230822

[79 rows x 1 columns]

The first approach I tried was a default without encoding the date. This achieved a test score of 0.79.
The second approach was encoding time as a number from the split date to see if the date would help predict the target. The test score was 0.78, which was worse than the score achieved with the default.
The third approach was to add one-hot encoding of the month for the Date field since individual days were not useful in predicting the target. This method achieved a test score of 0.80.

There is a lot of variability on a day to day basis, so it makes sense that the second approach yielded the worst results. By aggregating days into months, as in the third approach, the data became normalized, enabling more general patterns to be extracted by the model. Further, the seasonality of avocados is better modelled on a month-to-month basis rather than day-to-day. Thus, it makes sense that the third approach had the best test score.

<!-- END QUESTION -->

<br><br><br><br>

## Exercise 2: very short answer questions

Each question is worth 2 points.

<!-- BEGIN QUESTION -->

### 2.1 Time series

_Points:_ 4

The following questions pertain to Lecture 19 on time series data:

1. Sometimes a time series has missing time points or, worse, time points that are unequally spaced in general. Give an example of a real world situation where the time series data would have unequally spaced time points.
2. In class we discussed two approaches to using temporal information: encoding the date as one or more features, and creating lagged versions of features. Which of these (one/other/both/neither) two approaches would struggle with unequally spaced time points? Briefly justify your answer.

<div class="alert alert-warning">

Solution_2.1
    
</div>

1. Environmental monitoring systems collect data on parameters such as air quality, water quality, and pollution levels at irregular intervals based on factors such as sensor locations, sampling frequency, and environmental events. For example, air quality sensors might record data more frequently in urban areas with high pollution levels or during events such as wildfires, while data collection might be less frequent in rural areas with lower pollution levels. As a result, the time series data would have unequally spaced time points, reflecting the variability in environmental conditions and monitoring practices.

2. Both approaches would struggle with unequally spaced time points. Encoding the data as features assumes that the interval between time points is consistent, so when the time points are irregularly spaced the model may develop inaccurate interpretatations of temporal patterns. Creating lagged version of features involves creating lagged versions of features relative to a fixed lag time. If time points are unevenly spaced, the lagged version may not align correctly with the corresponding target observations, leading to inaccuracies in the model.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.2 Survival analysis

_Points:_ 6

The following questions pertain to Lecture 20 on survival analysis. We'll consider the use case of customer churn analysis.

1. What is the problem with simply labeling customers are "churned" or "not churned" and using standard supervised learning techniques?
2. Consider customer A who just joined last week vs. customer B who has been with the service for a year. Who do you expect will leave the service first: probably customer A, probably customer B, or we don't have enough information to answer?
3. If a customer's survival function is almost flat during a certain period, how do we interpret that?

<div class="alert alert-warning">

Solution_2.2
    
</div>

_Type your answer here, replacing this text._

<!-- END QUESTION -->

<br><br>

**PLEASE READ BEFORE YOU SUBMIT:** When you are ready to submit your assignment do the following:

1. Run all cells in your notebook to make sure there are no errors by doing `Kernel -> Restart Kernel and Clear All Outputs` and then `Run -> Run All Cells`. 
2. Notebooks with cell execution numbers out of order or not starting from “1” will have marks deducted. Notebooks without the output displayed may not be graded at all (because we need to see the output in order to grade your work).
3. Upload the assignment using [PrairieLearn](https://ca.prairielearn.com/pl/course_instance/6697). Don't forget to rename your file `hw8_sol.ipynb`.

![](img/eva-well-done.png)